src/kudu/consensus/raft_consensus.cc

```diff
@@ -3297,6 +3335,280 @@ void RaftConsensus::SetLeaderDetectedCallback(LeaderDetectedCallback ldcb) {
+                                                  "Local UUID: $1. Requested UUID: $2",
+                                                  peer_uuid(), request->proxy_dest_uuid()));
+    LOG_WITH_PREFIX(WARNING) << s.ToString() << ": from " << context->requestor_string()
+                 << ": " << SecureShortDebugString(*request);
+    SetupErrorAndRespond(s, ServerErrorPB::WRONG_SERVER_UUID, response, context);
+    return;
+  }
+  if (request->dest_uuid() == peer_uuid()) {
+    LOG_WITH_PREFIX(WARNING) << "dest_uuid and proxy_dest_uuid are the same: "
+                             << request->proxy_dest_uuid() << ": "
+                             << request->ShortDebugString();
+    context->RespondFailure(Status::InvalidArgument("proxy and desination must be different"));
+    return;
+  }
+
+  // Construct the downstream request; copy the relevant fields from the
+  // proxied request.
+  ConsensusRequestPB downstream_request;
+  auto prevent_ops_deletion = MakeScopedCleanup([&]() {
+    // Prevent double-deletion of these requests.
+    downstream_request.mutable_ops()->ExtractSubrange(
+      /*start=*/ 0, /*num=*/ downstream_request.ops_size(), /*elements=*/ nullptr);
+  });
+  downstream_request.set_dest_uuid(request->dest_uuid());
+  downstream_request.set_tablet_id(request->tablet_id());
+  downstream_request.set_caller_uuid(request->caller_uuid());
+  downstream_request.set_caller_term(request->caller_term());
+
+  if (request->has_preceding_id()) {
+    *downstream_request.mutable_preceding_id() = request->preceding_id();
+  }
+  if (request->has_committed_index()) {
+    downstream_request.set_committed_index(request->committed_index());
+  }
+  if (request->has_all_replicated_index()) {
+    downstream_request.set_all_replicated_index(request->all_replicated_index());
+  }
+  if (request->has_safe_timestamp()) {
+    downstream_request.set_safe_timestamp(request->safe_timestamp());
+  }
+  if (request->has_last_idx_appended_to_leader()) {
+    downstream_request.set_last_idx_appended_to_leader(request->last_idx_appended_to_leader());
+  }
+
+  downstream_request.set_proxy_caller_uuid(peer_uuid());
+
+  string next_uuid = request->dest_uuid();
+  if (FLAGS_raft_enable_multi_hop_proxy_routing) {
+    RET_RESPOND_ERROR_NOT_OK(routing_table_->NextHop(peer_uuid(), request->dest_uuid(), &next_uuid));
+  }
+
+  if (request->dest_uuid() != next_uuid) {
+    // Multi-hop proxy request.
+    downstream_request.set_proxy_dest_uuid(next_uuid);
+    // Forward the existing PROXY_OP ops.
+    for (int i = 0; i < request->ops_size(); i++) {
+      *downstream_request.add_ops() = request->ops(i);
+    }
+    prevent_ops_deletion.cancel(); // The ops we copy here are not pre-allocated.
+  } else {
+
+    // Reconstitute proxied events from the local cache.
+    // If the cache does not have all events, we retry up until the specified
+    // retry timeout.
+    // TODO(mpercy): Switch this from polling to event-triggered.
+
```

Comment:
i think this TODO is critical. The approach of simple polling will likely have issues.
My fear is it will increase cpu and waste lots of attempts to proxy messages to downstream nodes

Reply:
Yeah I will follow up to improve this. It’s perf work.